storage/storage_image.go

```diff
@@ -476,14 +541,167 @@ func (s *storageImageDestination) getConfigBlob(info types.BlobInfo) ([]byte, er
 	return nil, errors.New("blob not found")
 }
 
-func (s *storageImageDestination) Commit() error {
-	// Find the list of layer blobs.
-	if len(s.manifest) == 0 {
+// commitDataBlobs adds the non-layer blobs as data items to an image.   Since we only share layers,
+// they should all be in files, so we just need to screen out the ones that actually are layers to
+// get the list of non-layers.
+func (s *storageImageDestination) commitDataBlobs(imageID string) error {
+	dataBlobs := make(map[digest.Digest]struct{})
+	for blob := range s.filenames {
+		dataBlobs[blob] = struct{}{}
+	}
+	for layerDigest := range s.allLayerSums {
+		delete(dataBlobs, layerDigest)
+	}
+	for blob := range dataBlobs {
+		v, err := ioutil.ReadFile(s.filenames[blob])
+		if err != nil {
+			return errors.Wrapf(err, "error copying non-layer blob %q to image", blob)
+		}
+		if err := s.imageRef.transport.store.SetImageBigData(imageID, blob.String(), v); err != nil {
+			if _, err2 := s.imageRef.transport.store.DeleteImage(imageID, true); err2 != nil {
+				logrus.Debugf("error deleting incomplete image %q: %v", imageID, err2)
+			}
+			logrus.Debugf("error saving big data %q for image %q: %v", blob.String(), imageID, err)
+			return errors.Wrapf(err, "error saving big data %q for image %q", blob.String(), imageID)
+		}
+	}
+	return nil
+}
+
+// commitName adds the name from the reference to the specified image.  If the instance digest is
+// supplied, a tag or digest from the reference is replaced with the digest.
+func (s *storageImageDestination) commitName(imageID string, oldNames []string, instanceDigest *digest.Digest) error {
+	if name := s.imageRef.DockerReference(); len(oldNames) > 0 || name != nil {
+		var err error
+		if instanceDigest != nil {
+			name, err = reference.WithDigest(reference.TrimNamed(name), *instanceDigest)
+			if err != nil {
+				return errors.Wrapf(err, "error computing name for image %q", imageID)
+			}
+		}
+		names := []string{}
+		if name != nil {
+			names = append(names, verboseName(name))
+		}
+		if len(oldNames) > 0 {
+			names = append(names, oldNames...)
```

Comment:
Can this add `name` redundantly even if it is already present?

Reply:
In this `names` list, yes, but the storage library deduplicates the lists that are passed to all of its `SetNames()` methods.  I can add logic to also do that here if you prefer.

Reply:
No need to compute the same thing twice.

Maybe add a comment explaining that `SetNames` does this? OTOH this is no different from any other case where the reader is expected to understand how the called interface works :)

Reply:
> the reader

i.e. lazy me